src/dag-node.js

```diff
@@ -2,102 +2,134 @@
 
 const stable = require('stable')
 const mh = require('multihashes')
-const waterfall = require('async/waterfall')
+const series = require('async/series')
+const assert = require('assert')
 
 const DAGLink = require('./dag-link')
 
+function immutableError () {
+  throw new Error('Immutable property')
+}
+
 class DAGNode {
-  constructor (data, links) {
-    this._cached = {}
-    this._updated = false
-
-    this.data = data
-    this.links = []
-
-    // validate links
-    if (links) {
-      links.forEach((l) => {
-        if (l.constructor && l.constructor.name === 'DAGLink') {
-          this.links.push(l)
-        } else {
-          this.links.push(
-            new DAGLink(l.Name, l.Size, l.Hash)
-          )
-        }
-      })
+  constructor (data, links, serialized, size, multihash, json) {
+    assert(serialized, 'DAGNode needs its serialized format')
+    assert(multihash, 'DAGNode needs its multihash')
+    assert(json, 'DAGNode needs its json representation')
 
-      stable.inplace(this.links, util.linkSort)
+    if (typeof multihash === 'string') {
+      multihash = mh.fromB58String(multihash)
     }
-  }
 
-  /*
-   * addNodeLink - adds a DAGLink to this node that points
-   * to node by a name
-   */
-  addNodeLink (name, node, callback) {
-    if (typeof name !== 'string') {
-      throw new Error('first argument must be link name')
-    }
-    this.makeLink(node, (err, link) => {
-      if (err) {
-        return callback(err)
-      }
-      link.name = name
-      this.addRawLink(link)
-      callback()
+    Object.defineProperty(this, 'data', {
+      get () {
+        return data
+      },
+      set: immutableError
```

Comment:
well, that is what I did, but defined the function above so that I could reuse it